docs/rfcs/functional-composition-pattern.md

```diff
@@ -0,0 +1,311 @@
+# Functional Composition Pattern
+
+## Overview
+
+This codebase uses a **Functional Composition** pattern for its core
+interfaces. This pattern decomposes the individual methods from
+interface types into corresponding function types, then recomposes
+them into a concrete implementation type. This approach provides
+flexibility, testability, and safe interface evolution for the
+OpenTelemetry Collector.
+
+When an interface type is exported for users outside of this
+repository, the type MUST follow these guidelines. Interface types
+exposed from internal packages may opt-out of this recommendation.
+
+For every method in the public interface, a corresponding `type
+<Method>Func func(...) ...` declaration in the same package will
+exist, having the matching signature.
+
+For every interface type, there is a corresponding functional
+constructor `func New<Type>(<Method1>Func, <Method2>Func, ...) Type`
+in the same package for constructing a functional composition of
+interface methods.
+
+Interface stability for exported interface types is our primary
+objective. The Functional Composition pattern supports safe interface
+evolution, first by "sealing" the type with an unexported interface
+method. This means all implementations of an interface must embed or
+use constructors provided in the package.
+
+These "sealed concrete" implementation objects support adding new
+methods in future releases, _without changing the major version
+number_, because public interface types are always provided through a
+package-provided implementation.
+
+As a key requirement, every function must have a simple "no-op"
+implementation corresponding with the zero value of the
+`<Method>Func`. The expression `New<Type>(nil, nil, ...)` is the empty
+implementation for each type.
```

Comment:
The no-op is true, but to be able to extend the `New<Type>(nil, nil, ...)` we decided to split the functions into required and optional (which by the definition of the interface is fine to be no-op in lots of cases), and use an optional pattern. This may not be the best tradeoff, but allows us to extend the optional functions and not the required, which is ok since any new func/method added to any interface needs to have a default implementation.

Reply:
Got it. I had described (as you saw) the potential to start with a non-option-taking constructor and add a `New<Type>WithOptions` later.! I believe you are saying that constructors should or must support the functional option pattern even when there are no options, from the start. This allows adding options in the future, always without new constructors.

Reply:
> believe you are saying that constructors should or must support the functional option pattern even when there are no options, from the start. 

Correct